flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java

```diff
@@ -2090,6 +2092,25 @@ public StreamGraph getStreamGraph(String jobName) {
     @Internal
     public StreamGraph getStreamGraph(String jobName, boolean clearTransformations) {
         StreamGraph streamGraph = getStreamGraphGenerator().setJobName(jobName).generate();
+
+        // There might be a resource deadlock when applying fine-grained resource management in
+        // batch jobs with PIPELINE edges. Users need to trigger the
+        // fine-grained.shuffle-mode.all-blocking to convert all edges to BLOCKING before we fix
+        // that issue in FLINK-20865.
+        if (configuration.get(ExecutionOptions.RUNTIME_MODE) == RuntimeExecutionMode.BATCH
+                && !slotSharingGroupResources.isEmpty()) {
+            if (configuration.get(ClusterOptions.FINE_GRAINED_SHUFFLE_MODE_ALL_BLOCKING)) {
+                streamGraph.setGlobalDataExchangeMode(GlobalDataExchangeMode.ALL_EDGES_BLOCKING);
+            } else {
+                throw new IllegalStateException(
+                        "There might be a resource deadlock, which will be fixed in FLINK-20865, when "
+                                + "applying fine-grained resource management in batch jobs with PIPELINE edges. "
+                                + "As a temporary workaround, please configure the "
+                                + ClusterOptions.FINE_GRAINED_SHUFFLE_MODE_ALL_BLOCKING.key()
+                                + " to true, with which Flink will convert those PIPELINE edges to BLOCKING.");
+            }
```

Comment:
1. I would say "as described in FLINK-20865" rather than "will be fixed in FLINK-20865". We do not know for sure that the problem will be fixed in FLINK-20865, thus should not commit that to users.
2. It is important to state that changing from PIPELINE to BLOCKING edges might affect the performance. Since `FINE_GRAINED_SHUFFLE_MODE_ALL_BLOCKING` is `ExcludeFromDocumentation`, its description will not be seen by users.